modules/stochastic_tools/src/surrogates/LibtorchSimpleNNTrainer.C

```diff
@@ -0,0 +1,229 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "LibtorchSimpleNNTrainer.h"
+#include "Sampler.h"
+
+registerMooseObject("StochasticToolsApp", LibtorchSimpleNNTrainer);
+
+InputParameters
+LibtorchSimpleNNTrainer::validParams()
+{
+  InputParameters params = SurrogateTrainer::validParams();
+
+  params.addClassDescription("Trains a simple neural network using libtorch.");
+
+  MooseEnum data_type("real=0 vector_real=1", "real");
+  params.addRequiredParam<ReporterName>(
+      "response",
+      "Reporter value of response results, can be vpp with <vpp_name>/<vector_name> or sampler "
+      "column with 'sampler/col_<index>'.");
+  params.addParam<MooseEnum>("response_type", data_type, "Response data type.");
+  params.addParam<unsigned int>("no_batches", 1, "Number of batches.");
+  params.addParam<unsigned int>("no_epochs", 1, "Number of epochs.");
+  params.addParam<unsigned int>("no_hidden_layers", 0, "Number of hidden layers.");
+  params.addParam<std::vector<unsigned int>>(
+      "no_neurons_per_layer", std::vector<unsigned int>(), "Number of neurons per layer.");
+  params.addParam<std::string>(
+      "filename", "net.pt", "Filename used to output the neural net parameters.");
+  params.addParam<bool>("read_from_file",
+                        false,
+                        "Switch to allow reading old trained neural nets for further training.");
+  params.addParam<Real>("learning_rate", 0.001, "Learning rate (relaxation).");
+  params.addParam<unsigned int>(
+      "seed", 11, "Random number generator seed for stochastic optimizers.");
+
+  return params;
+}
+
+LibtorchSimpleNNTrainer::LibtorchSimpleNNTrainer(const InputParameters & parameters)
+  : SurrogateTrainer(parameters),
+    _sampler_row(getSamplerData()),
+    _response(getTrainingData<Real>(getParam<ReporterName>("response"))),
+    _no_batches(getParam<unsigned int>("no_batches")),
+    _no_epocs(getParam<unsigned int>("no_epochs")),
+    _no_hidden_layers(declareModelData<unsigned int>("no_hidden_layers")),
```

Comment:
@grmnptr Just a reminder before this gets merged: the `declareModelData` here does not seem to pass the correct parameter values (for both `_no_hidden_layers` and `_no_neurons_per_layer`) to the model. This caused some trouble for me while calling `forward`.

Reply:
Thanks for the reminder! I'll check it but we will try to keep it because it grants restarting capabilities between the trainer and surrogate objects.

Reply:
I'll get back to you here in this conversation once I diagnosed the behavior.

Reply:
I tested the trainer and I think it is working as intended. The declareModel just creates a container at the beginning which can be filled later/at the initialization. I assigned values the the model data a little bit later in the contructor in this version which is probably why they looked empty right at the beginning. I shortened the code and now we assign values right at initialization.